include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h

```diff
@@ -0,0 +1,187 @@
+
+/// Find the original definition of a potentially copied value.
+///
+/// This use-def walk must be consistent with the def-use walks performed within
+/// the CanonicalOSSALifetime implementation.
+inline SILValue getCanonicalCopiedDef(SILValue v) {
+  while (true) {
+    if (auto *copy = dyn_cast<CopyValueInst>(v)) {
+      v = copy->getOperand();
+      continue;
+    }
+    return v;
+  }
+}
+
+/// Information about consumes on the extended-lifetime boundary. Consuming uses
+/// within the lifetime are not included--they will consume a copy after
+/// rewriting.
+///
+/// This result remains valid during copy rewriting. The only instructions
+/// referenced it contains are consumes that cannot be deleted.
+class CanonicalOSSAConsumeInfo {
+  // Map blocks on the lifetime boundary to the last consuming instruction.
+  llvm::SmallDenseMap<SILBasicBlock *, SILInstruction *, 4> finalBlockConsumes;
+
+public:
+  bool empty() const { return finalBlockConsumes.empty(); }
+
+  void clear() { finalBlockConsumes.clear(); }
+
+  void recordFinalConsume(SILInstruction *inst) {
+    assert(!finalBlockConsumes.count(inst->getParent()));
+    finalBlockConsumes[inst->getParent()] = inst;
+  }
+
+  // Return true if this instruction is marked as a final consume point of the
+  // current def's live range. A consuming instruction can only be claimed once
+  // because instructions like `tuple` can consume the same value via multiple
+  // operands.
+  bool claimConsume(SILInstruction *inst) {
+    auto destroyPos = finalBlockConsumes.find(inst->getParent());
+    if (destroyPos != finalBlockConsumes.end() && destroyPos->second == inst) {
+      finalBlockConsumes.erase(destroyPos);
+      return true;
+    }
+    return false;
+  }
+
+  SWIFT_ASSERT_ONLY_DECL(void dump() const LLVM_ATTRIBUTE_USED);
+};
+
+/// Canonicalize OSSA lifetimes.
+///
+/// Allows the allocation of analysis state to be reused across calls to
+/// canonicalizeValueLifetime().
+struct CanonicalOSSALifetime {
+  // Current copied def for which this state describes the liveness.
+  SILValue currDef;
+
+  // Cumulatively, have any instructions been modified by canonicalization?
+  bool changed = false;
+
+  // Pruned liveness for the extended live range including copies. For this
+  // purpose, only consuming instructions are considered "lifetime
+  // ending". end_borrows do not end a liverange that may include owned copies.
+  PrunedLiveness liveness;
+
+  // Original points in the CFG where the current value's lifetime ends. This
+  // includes any point in which the value is consumed or destroyed. For
+  // guaranteed values, it also includes points where the borrow scope
+  // ends. A backward walk from these blocks must discover all uses on paths
+  // that lead to a return or throw.
+  //
+  // These blocks are not necessarily in the pruned live blocks since
+  // pruned liveness does not consider destroy_values.
+  llvm::SmallSetVector<SILBasicBlock *, 8> lifetimeEndBlocks;
```

Comment:
You don't need llvm:: for this. We already import it into the swift namespace in Basic/LLVM.h.